firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/internal/ProxyAnalyticsConnector.java

```diff
@@ -0,0 +1,177 @@
+ */
+public class ProxyAnalyticsConnector implements AnalyticsConnector {
+  private volatile Object instance;
+
+  public ProxyAnalyticsConnector(Deferred<AnalyticsConnector> analyticsConnector) {
+    instance = analyticsConnector;
+    analyticsConnector.whenAvailable(connectorProvider -> instance = connectorProvider.get());
+  }
+
+  private AnalyticsConnector safeGet() {
+    Object result = instance;
+    if (result instanceof AnalyticsConnector) {
+      return (AnalyticsConnector) result;
+    }
+    return null;
+  }
+
+  @Override
+  public void logEvent(@NonNull String s, @NonNull String s1, @NonNull Bundle bundle) {
+    AnalyticsConnector connector = safeGet();
+    if (connector != null) {
+      connector.logEvent(s, s1, bundle);
+    }
+  }
+
+  @Override
+  public void setUserProperty(@NonNull String s, @NonNull String s1, @NonNull Object o) {
+    AnalyticsConnector connector = safeGet();
+    if (connector != null) {
+      connector.setUserProperty(s, s1, o);
+    }
+  }
+
+  // Not implemented since it's not used by fiam
+  @NonNull
+  @Override
+  public Map<String, Object> getUserProperties(boolean b) {
+    return Collections.emptyMap();
+  }
+
+  @NonNull
+  @Override
+  public AnalyticsConnectorHandle registerAnalyticsConnectorListener(
+      @NonNull String s, @NonNull AnalyticsConnectorListener analyticsConnectorListener) {
+    Object result = instance;
+    if (result instanceof AnalyticsConnector) {
+      return ((AnalyticsConnector) result)
+          .registerAnalyticsConnectorListener(s, analyticsConnectorListener);
+    }
+    @SuppressWarnings("unchecked")
+    Deferred<AnalyticsConnector> deferred = (Deferred<AnalyticsConnector>) result;
+    return new ProxyAnalyticsConnectorHandle(s, analyticsConnectorListener, deferred);
+  }
+
+  // Not implemented since it's not used by fiam.
+  @Override
+  public void setConditionalUserProperty(
+      @NonNull ConditionalUserProperty conditionalUserProperty) {}
+
+  // Not implemented since it's not used by fiam.
+  @Override
+  public void clearConditionalUserProperty(
+      @NonNull String s, @Nullable String s1, @Nullable Bundle bundle) {}
+
+  // Not implemented since it's not used by fiam.
+  @NonNull
+  @Override
+  public List<ConditionalUserProperty> getConditionalUserProperties(
+      @NonNull String s, @Nullable String s1) {
+    return Collections.emptyList();
+  }
+
+  // Not implemented since it's not used by fiam.
+  @Override
+  public int getMaxUserProperties(@NonNull String s) {
+    return 0;
+  }
+
+  private static class ProxyAnalyticsConnectorHandle implements AnalyticsConnectorHandle {
+    private static final Object UNREGISTERED = new Object();
+    private Set<String> eventNames = new HashSet<>();
+    private Object instance;
+
+    private ProxyAnalyticsConnectorHandle(
+        String s,
+        AnalyticsConnectorListener listener,
+        Deferred<AnalyticsConnector> analyticsConnector) {
+      analyticsConnector.whenAvailable(
```

Comment:
Is there an ordering the whenAvaiable callbacks are executed? Should we be reasoning about that?